### PR TITLE
Reduce TaskAlignedAssigner OOM recovery cost

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

- retry failed TaskAlignedAssigner batches per image on GPU instead of moving the full assignment to CPU
- warn once per training while reporting the assignment shape and confirming that the model forward batch is unchanged
- trim trailing GT padding, preallocate retry outputs, retain candidate masks as int8, and reuse dense metric buffers
- compute box metrics only for valid anchor/GT pairs

## xView benchmark

RTX PRO 6000 Blackwell, CUDA AMP, cached xView batch 128 at 640 px: 75,357 objects, 4,967 max objects/image, 8,400 anchors.

| Path | Time | Peak allocated |
| --- | ---: | ---: |
| Existing GPU OOM -> CPU | 194.86 s | 85.62 GB |
| Existing per-image GPU | 0.442 s | 1.90 GB |
| Optimized automatic OOM -> per-image GPU | 0.699 s | 58.98 GB |
| Optimized direct per-image GPU | 0.376 s | 1.56 GB |

Warmed bounded chunks of 2/4/8/16/32 images were all slower and used more memory than single-image assignment.

CUDA AMP comparison against the existing whole-batch GPU path was bit-exact for all five outputs. The existing CPU fallback changed two foreground assignments on this batch because CUDA autocast FP32 `pow()` promotion does not follow tensors to CPU.

## Validation

- `uvx ruff check ultralytics/utils/tal.py`
- `uvx ruff format --check ultralytics/utils/tal.py ultralytics/utils/torch_utils.py`
- detection and OBB one-epoch CPU smoke training
- detection and rotated assigner equivalence with default and secondary top-k
- CUDA AMP bit-exact comparison against `main`
- forced repeated OOM recovery emits one warning, restores assigner state, and propagates a per-image OOM to Trainer

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

TaskAlignedAssigner now recovers from GPU out-of-memory errors by retrying assignment one image at a time on the GPU, while reducing temporary memory use during metric computation.

### 📊 Key Changes

- Replaced full-batch CPU fallback with per-image GPU retries that trim trailing ground-truth padding, preallocate batched outputs, and restore assigner state afterward.
- Added a one-time warning containing the assignment batch size and maximum object count, explicitly noting that the model forward batch size remains unchanged.
- Reduced assignment memory use by computing box metrics only for valid anchor/ground-truth pairs, using in-place metric operations, and retaining candidate masks without conversion to the metric dtype.
- Simplified candidate-in-ground-truth masking and removed an unnecessary float conversion when resolving multi-ground-truth assignments.
- Updated the package version from `8.4.133` to `8.4.134` and adjusted operation profiling comments for the revised memory estimate.

### 🎯 Purpose & Impact

- GPU OOM recovery no longer moves the entire assignment operation and its inputs to the CPU; it retries each image independently on the GPU.
- Successful retries return outputs with the original training batch shape, while a per-image OOM remains available to propagate to the trainer.
- Repeated recovery warnings are suppressed after the first warning per assigner instance.
- Assignment state is restored after retry processing, including when a retry raises an error.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/tal.py [hunk 1]
```diff
@@ -99,11 +97,32 @@ def forward(self, pd_scores, pd_bboxes, anc_points, gt_labels, gt_bboxes, mask_g
...
+            .mul(torch.arange(1, n_max_boxes + 1, device=mask_gt.device))
+            .amax(1)
+            .clamp_(min=1)
+            .tolist()
+        )
+        self.bs = 1
+        results = []
+        for i, self.n_max_boxes in enumerate(last_gt_idx):
+            results.append(
+                self._forward(
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/tal.py]: ⚠️ **HIGH**: The previous implementation retained a CPU fallback for any assignment OOM, but these per-image retries run after the `try/except`. If the batch has one image, or an individual image still exceeds GPU memory, this `_forward()` raises an uncaught OOM and training aborts instead of falling back to CPU as before. Catch retry OOMs and preserve a CPU fallback for the images that do not fit on GPU.

File: ultralytics/utils/tal.py [hunk 2]
```diff
@@ -99,11 +97,32 @@ def forward(self, pd_scores, pd_bboxes, anc_points, gt_labels, gt_bboxes, mask_g
...
+                    pd_scores[i : i + 1],
+                    pd_bboxes[i : i + 1],
+                    anc_points,
+                    gt_labels[i : i + 1, : self.n_max_boxes],
+                    gt_bboxes[i : i + 1, : self.n_max_boxes],
+                    mask_gt[i : i + 1, : self.n_max_boxes],
+                )
+            )
+        self.bs, self.n_max_boxes = bs, n_max_boxes
+        return tuple(torch.cat(x, 0) for x in zip(*results))
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/tal.py]: 💡 **MEDIUM**: `results` keeps every per-image output alive while `torch.cat` allocates a second full-batch copy. For large batches or class counts, this extra output-sized peak can itself raise OOM after all per-image assignments succeeded, and this exception is also outside the recovery handler. Preallocate the final outputs and copy each result into its batch slice, or otherwise release per-image tensors before concatenation.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the TaskAlignedAssigner OOM recovery, sparse metric computation, mask dtype changes, and profiling estimate. The optimized assignment path is otherwise behaviorally consistent, but the new recovery path can lose the prior CPU fallback and can re-trigger OOM during final concatenation.

<details><summary>💬 Posted 2 inline comments</summary>

- ⚠️ **HIGH** `ultralytics/utils/tal.py:115` The previous implementation retained a CPU fallback for any assignment OOM, but these per-image retries run after the `try/except`. If the batch has one image, or an individual image still exceeds GPU memory, this `_forward()` raises an uncaught OOM and training aborts instead of falling back to CPU as before. Catch retry OOMs and preserve a CPU fallback for the images that do not fit on GPU.
- 💡 **MEDIUM** `ultralytics/utils/tal.py:125` `results` keeps every per-image output alive while `torch.cat` allocates a second full-batch copy. For large batches or class counts, this extra output-sized peak can itself raise OOM after all per-image assignments succeeded, and this exception is also outside the recovery handler. Preallocate the final outputs and copy each result into its batch slice, or otherwise release per-image tensors before concatenation.
</details>

File: ultralytics/utils/tal.py (hunk 2 above)
Comment (glenn-jocher) [reply ultralytics/utils/tal.py]: Fixed in 0a01b9174: the fallback now allocates each full-batch output once from the first per-image result and fills its batch slices, eliminating the output-sized concatenation peak.

File: ultralytics/utils/tal.py (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics/utils/tal.py]: This is intentional. TaskAlignedAssigner now owns only the bounded GPU retry; if one image still cannot fit, the OOM propagates to Trainer, which owns batch-size recovery. Restoring the CPU path would reintroduce the 195 s stall and changes AMP assignment results relative to GPU. The temporary assigner shape state is restored in a finally block before propagation.

**Comment by UltralyticsAssistant** [comment]
🎉 Thank you, @glenn-jocher, for merging PR [#25990](https://github.com/ultralytics/ultralytics/pull/25990)!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This release brings together targeted memory reductions and per-image GPU retry handling to improve TaskAlignedAssigner OOM recovery while preserving the original training batch shape and restoring assigner state safely.
